client/finality-grandpa/src/tests.rs

```diff
@@ -1284,105 +1283,99 @@ fn voter_persists_its_votes() {
-							target_hash: block_30_hash,
-						};
-
-						round_tx.lock().start_send(finality_grandpa::Message::Prevote(prevote)).unwrap();
-						Ok(())
-					}).map_err(|_| panic!()))
-
-			} else if state.compare_and_swap(1, 2, Ordering::SeqCst) == 1 {
-				// the next message we receive should be our own prevote
-				let prevote = match signed.message {
-					finality_grandpa::Message::Prevote(prevote) => prevote,
-					_ => panic!("We should receive our own prevote."),
-				};
-
-				// targeting block 30
-				assert!(prevote.target_number == 30);
-
-				// after alice restarts it should send its previous prevote
-				// therefore we won't ever receive it again since it will be a
-				// known message on the gossip layer
-
-				future::Either::B(future::ok(()))
-
-			} else if state.compare_and_swap(2, 3, Ordering::SeqCst) == 2 {
-				// we then receive a precommit from alice for block 15
-				// even though we casted a prevote for block 30
-				let precommit = match signed.message {
-					finality_grandpa::Message::Precommit(precommit) => precommit,
-					_ => panic!("voter should precommit."),
-				};
-
-				assert!(precommit.target_number == 15);
-
-				// signal exit
-				exit_tx.clone().lock().take().unwrap().send(()).unwrap();
-
-				future::Either::B(future::ok(()))
-
-			} else {
-				panic!()
+			let net2 = net.clone();
+			let net = net.clone();
+			let voter_tx = voter_tx.clone();
+			let round_tx = round_tx.clone();
+			let state = state.clone();
+			let exit_tx = exit_tx.clone();
+
+			async move {
+				if state.compare_and_swap(0, 1, Ordering::SeqCst) == 0 {
+					// the first message we receive should be a prevote from alice.
+					let prevote = match signed.message {
+						finality_grandpa::Message::Prevote(prevote) => prevote,
+						_ => panic!("voter should prevote."),
+					};
+
+					// its chain has 20 blocks and the voter targets 3/4 of the
+					// unfinalized chain, so the vote should be for block 15
+					assert!(prevote.target_number == 15);
+
+					// we push 20 more blocks to alice's chain
+					net.lock().peer(0).push_blocks(20, false);
+
+					let interval = futures::stream::unfold(Delay::new(Duration::from_millis(200)), |delay|
+						Box::pin(async move {
+							delay.await;
+							Some(((), Delay::new(Duration::from_millis(200))))
+						})
+					);
+
+					interval
+						.take_while(move |_| {
+							future::ready(net2.lock().peer(1).client().info().best_number != 40)
+						})
+						.for_each(|_| future::ready(()))
+						.await;
+
+					let block_30_hash =
+						net.lock().peer(0).client().as_full().unwrap().hash(30).unwrap().unwrap();
+
+					// we restart alice's voter
+					voter_tx.unbounded_send(()).unwrap();
+
+					// and we push our own prevote for block 30
+					let prevote = finality_grandpa::Prevote {
+						target_number: 30,
+						target_hash: block_30_hash,
+					};
+
+					Pin::new(&mut *round_tx.lock()).start_send(finality_grandpa::Message::Prevote(prevote)).unwrap();
```

Comment:
One should first call `poll_ready` before `start_send` on ` `Sink`. As a convenience function, given that we are within an `async` block, one can call `SinkExt::send`.

(I don't think this was introduced through this pull request, but the pull request enables the easy fix via `SinkExt::send`. Let me know if you want me to do this as a follow up.)

Reply:
We actually can't use `SinkExt::send`, because that's that'd mean awaiting a function that borrows a mutex and ends up with: `*mut () cannot be sent between threads safely`.